lib/screens/v2/candidate_generator/dup/departures.ex

```diff
@@ -308,17 +382,159 @@ defmodule Screens.V2.CandidateGenerator.Dup.Departures do
-  # Otherwise, we only need the first route in the list of routes serving the stop.
-  defp get_primary_route_for_section([stop_id | _], route_ids, create_station_with_routes_map_fn) do
-    all_routes = create_station_with_routes_map_fn.(stop_id)
+  # If we are currently overnight, returns the first schedule of the day for each route_id and direction for each stop.
+  # Otherwise, return an empty list.
+  defp get_overnight_schedules_for_section(
+         stops_with_live_departures,
+         stop_ids,
+         routes_serving_section,
+         alert,
+         now,
+         fetch_schedules_fn,
+         fetch_vehicles_fn
+       )
+
+  # No predictions AND no active alerts for the section
+  defp get_overnight_schedules_for_section(
+         stops_with_live_departures,
+         params,
+         routes,
+         nil,
+         now,
+         fetch_schedules_fn,
+         _
+       ) do
+    {today_schedules, tomorrow_schedules} =
+      get_today_tomorrow_schedules(
+        Map.from_struct(params),
+        fetch_schedules_fn,
+        Util.get_service_day_tomorrow(now),
+        Enum.map(routes, & &1.id)
+      )
+
+    # Get schedules for each stop_id in config
+    today_schedules
+    |> Enum.map(&{&1.stop.id, &1.direction_id})
+    |> Enum.uniq()
+    |> Enum.reject(&(&1 in stops_with_live_departures))
+    |> Enum.map(fn {stop_id, direction_id} ->
+      # If now is before any of today's schedules or after any of tomorrow's (should never happen but just in case),
+      # we do not display overnight mode.
+
+      last_schedule_today =
+        Enum.find(
+          today_schedules,
+          struct(Schedule, departure_time: now),
+          &(&1.stop.id == stop_id and &1.direction_id == direction_id)
+        )
+
+      first_schedule_tomorrow =
+        Enum.find(
+          tomorrow_schedules,
+          struct(Schedule, departure_time: now),
+          &(&1.stop.id == stop_id and &1.direction_id == direction_id)
+        )
+
+      if DateTime.compare(now, last_schedule_today.departure_time) == :gt and
+           DateTime.compare(now, first_schedule_tomorrow.departure_time) == :lt do
+        %Departure{schedule: first_schedule_tomorrow}
+      end
+    end)
+    # Routes not in overnight mode will be nil. Can ignore those.
+    |> Enum.reject(&is_nil/1)
+    |> Enum.sort_by(fn %Departure{schedule: schedule} -> schedule.departure_time end)
+  end
+
+  defp get_overnight_schedules_for_section(
+         stops_with_live_departures,
+         params,
+         [%{type: type} | _] = routes,
+         alert,
+         now,
+         fetch_schedules_fn,
+         fetch_vehicles_fn
+       )
+       when type in [:subway, :light_rail] do
+    informed_route = get_section_route_from_alert(params.stop_ids, alert)
+    # If there are no vehicles operating on the route, assume we are overnight.
+    if fetch_vehicles_fn.(informed_route.id, nil) == [] do
```

Comment:
Have you tested this out with an alert yet? For me, this fails because the output of `informed_route` is a string, and the dot access (`.id`) requires an atom. Alternatively, you could make it `informed_route[id]`. I think that'll work

Reply:
That `.id` should not have been there. I must have added that without thinking after doing my testing. The string returned is always a `route_id` (or `nil` which is why I added `not is_nil(informed_route)` to the conditional). That can then be used in the vehicle fetch.